### PR TITLE
Added admin/edit/<food_name> path

### PR DESCRIPTION
**Pull request by fernandalavalle**
This change is [Reviewable](https://reviewable.io/reviews/mtlynch/isitketo_webapp/65)

### CODE REVIEW

File: app/admin.py [hunk 1]
```diff
@@ -6,15 +8,17 @@
 app = flask.Flask(__name__)
 
 
-@app.route('/api/admin/add', methods=['POST'])
-def api_add_food():
+def _sanitize_food_name(food_name):
+    return re.sub(r'[^a-zA-Z\-\.\s&*\+:0-9]', '', food_name)
```
Comment (mtlynch) [inline app/admin.py]: Can we move this to a separate module and add unit tests? This is the kind of thing where it's easy to forget what characters we care about stripping and what we want to preserve.

Take a look at the [KetoHub ingredient parsing unit tests](https://github.com/mtlynch/ketohub_offline_scraper/blob/master/tests/test_ingredients.py) for an example structure of unit tests and some edge cases. Those are all ingredients so we can assume we're not going to get the "1 cups" part, but the brand and food names are in scope.

In particular, I think the current implementation would strip some Unicode characters that are actually okay, like:

>Rao’s marinara

Which is actually represented using a Unicode apostrophe:

```
u'Rao\u2019s marinara'
```

Also keep in mind that we don't have to be super aggressive about sanitizing against a malicious user because we assume the person editing is an admin.

File: app/templates/food_form.html [hunk 2]
```diff
@@ -0,0 +1,62 @@
+{% include 'header.html' %}
+
```
Comment (mtlynch) [inline app/templates/food_form.html]: Can we replace tabs with whitespaces in this file? The mix is causing the indenting to look weird on Github.

File: app/admin.py [hunk 3]
```diff
@@ -33,4 +39,28 @@ def api_add_food():
...
+def api_edit_food(food_name):
+    f = _make_food_from_form()
+    f = food.update(food_name, f)
+    return flask.redirect('/%s' % f.key.string_id())
+
+
+@app.route('/admin/edit/<food_name>')
+def edit_food(food_name):
+    user = users.get_current_user()
+    food_name = sanitizer.sanitize_food_path(food_name)
```
Comment (mtlynch) [inline app/admin.py]: One thing I've been more conscious of since learning Go is "stutter" naming:

>**Avoid stutter**. Since client code uses the package name as a prefix when referring to the package contents, the names for those contents need not repeat the package name. The HTTP server provided by the http package is called Server, not HTTPServer. Client code refers to this type as http.Server, so there is no ambiguity. 

https://blog.golang.org/package-names

We could simplify this by changing the module to `sanitize` and then just make the function `food_path`, so the semantics would look like `sanitize.food_path` which is just as clear, but avoids redundancy.

File: app/admin.py [hunk 4]
```diff
@@ -33,4 +39,32 @@ def api_add_food():
...
+def api_edit_food(food_name):
+    f = _make_food_from_form()
+    f = food.update(food_name, f)
+    return flask.redirect('/%s' % f.key.string_id())
+
+
+@app.route('/admin/edit/<food_name>')
+def edit_food(food_name):
+    print "got to edit food path"
+    print food.name_to_key(food_name).get()
```
Comment (mtlynch) [inline app/admin.py]: stray debug lines

File: app/sanitizer.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import re
+
+
+def food_name(food_name):
```
Comment (mtlynch) [inline app/sanitizer.py]: I think we should just inline this function directly in `name_to_key` and then unit test `name_to_key`. It's too confusing that we're doing some of the transformation here and some of the transformation there.

File: app/sanitizer.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    Replaces unicode characters to corresponding ASCII characters. This allows unicode to be mapped
+    as expected to comparable ASCII names.
+
+    Ex: u'jalapeno\u's -> jalapenos
+
+    Replaces characters that are not legal in file paths by GCS or Linux/Windows. Does not affect
+    how food names will appear to the user.
+
+    """
```
Comment (mtlynch) [inline app/sanitizer.py]: Missing args section.

File: app/sanitizer.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+def food_name(food_name):
+    """Sanitizes the food keys in the admin/edit/<food_name> path.
+
+    Replaces unicode characters to corresponding ASCII characters. This allows unicode to be mapped
+    as expected to comparable ASCII names.
+
+    Ex: u'jalapeno\u's -> jalapenos
+
+    Replaces characters that are not legal in file paths by GCS or Linux/Windows. Does not affect
+    how food names will appear to the user.
```
Comment (mtlynch) [inline app/sanitizer.py]: This is really the headline, but it's buried at the end of the docstring.

This is a function's purpose is to convert food names to keys. In order to do that, it needs to make the name satisfy the requirements of keys:

* Keys act as the canonical form of a food name, so that if the user types `jalapeños` or `jalapenos`, they both map to the same key
* Keys contain only characters that are valid in filenames on GCS, Linux, and Windows

We can then go on to say what particular transformations we're making, but the important part is why we're making these transformations.

File: app/sanitizer.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import re
+
+
+def food_name(food_name):
+    """Sanitizes the food keys in the admin/edit/<food_name> path.
```
Comment (mtlynch) [inline app/sanitizer.py]: It's generally an anti-pattern to talk about the caller in the docstring. The purpose of the function should be independent of who's calling it.

File: app/sanitizer.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    cleaner = re.sub(u'\u00f1', 'n', food_name)
+    # e with an accent to e.
+    cleaner = re.sub(u'[\u00e8-\u00eb]', 'e', cleaner)
+    # Gets rid of Unicode and ASCII apostrophes.
+    cleaner = re.sub(u'\u2019', '', cleaner)
+    cleaner = re.sub(r'\'', '', cleaner)
+    cleaner = re.sub('&', '-and-', cleaner)
+    # Get rid of ".
+    cleaner = re.sub(u'[\u201c-\u201d]', '', cleaner)
+    # Unicode dashes to spaces that will then become ASCII dashes below.
```
Comment (mtlynch) [inline app/sanitizer.py]: Suggest just converting to dashes so we don't have to depend on a particular ordering.

File: app/sanitizer.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import re
+
+
+def food_name(food_name):
+    """Sanitizes the food keys in the admin/edit/<food_name> path.
+
+    Replaces unicode characters to corresponding ASCII characters. This allows unicode to be mapped
+    as expected to comparable ASCII names.
+
+    Ex: u'jalapeno\u's -> jalapenos
```
Comment (mtlynch) [inline app/sanitizer.py]: I think this is supposed to be `jalape\u00f1os`

**Comment by mtlynch** [comment]
LGTM!
